test/mpi/multigrid/pgm.cpp

```diff
@@ -0,0 +1,125 @@
+
+    Pgm()
+        : size{8, 8}, mat_input{size, {{0, 0, 5},  {0, 1, -1}, {1, 0, -1},
+                                       {1, 1, 5},  {2, 2, 5},  {3, 3, 5},
+                                       {4, 4, 5},  {4, 6, -2}, {5, 5, 5},
+                                       {5, 7, -2}, {6, 4, -2}, {6, 6, 5},
+                                       {7, 5, -2}, {7, 7, 5},  {0, 2, -3},
+                                       {0, 4, 1},  {0, 5, 2},  {0, 6, 3},
+                                       {1, 3, -4}, {1, 5, 4},  {1, 6, 5},
+                                       {1, 7, 6},  {2, 0, -3}, {2, 5, -1},
+                                       {2, 6, -2}, {3, 1, -4}, {3, 7, -5},
+                                       {4, 0, 1},  {5, 0, 2},  {5, 1, 4},
+                                       {5, 2, -1}, {6, 0, 3},  {6, 1, 5},
+                                       {6, 2, -2}, {7, 1, 6},  {7, 3, -5}}}
+    {
+        row_part = Partition::build_from_contiguous(
+            exec, gko::array<global_index_type>(
+                      exec, I<global_index_type>{0, 2, 4, 8}));
+
+        dist_mat = dist_mtx_type::create(exec, comm);
+        dist_mat->read_distributed(mat_input, row_part.get());
+    }
+
+    void SetUp() override { ASSERT_EQ(comm.size(), 3); }
+
+    gko::dim<2> size;
+    std::shared_ptr<Partition> row_part;
+
+    gko::matrix_data<value_type, global_index_type> mat_input;
+
+    std::shared_ptr<dist_mtx_type> dist_mat;
+};
+// using One = ::testing::Types<std::tuple<float, gko::int32, gko::int32>>;
+using More =
+    ::testing::Types<std::tuple<float, gko::int32, gko::int32>,
+                     std::tuple<float, gko::int64, gko::int64>,
+                     std::tuple<std::complex<float>, gko::int64, gko::int64>,
+                     std::tuple<std::complex<float>, gko::int64, gko::int64>,
+                     std::tuple<double, gko::int32, gko::int32>,
+                     std::tuple<double, gko::int64, gko::int64>,
+                     std::tuple<std::complex<double>, gko::int64, gko::int64>,
+                     std::tuple<std::complex<double>, gko::int64, gko::int64>>;
+TYPED_TEST_SUITE(Pgm, gko::test::ValueLocalGlobalIndexTypes,
+                 TupleTypenameNameGenerator);
+
+
+TYPED_TEST(Pgm, CanGenerateFromDistributedMatrix)
+{
+    using pgm = typename TestFixture::pgm;
+    using value_type = typename TestFixture::value_type;
+    using dist_mtx_type = typename TestFixture::dist_mtx_type;
+    using local_matrix_type = typename TestFixture::local_matrix_type;
+    auto pgm_factory = pgm::build().on(this->exec);
+    auto rank = this->comm.rank();
+    I<I<value_type>> res_local[] = {{{8}}, {{5, 0}, {0, 5}}, {{6, 0}, {0, 6}}};
+    // the rank 2 part of non local matrix of rank 1 are reordered.
+    // [0 -1 -2 0], 1st and 3rd are aggregated to the first group but the rest
+    // are aggregated to the second group. Thus, the aggregated result should be
+    // [-2 -1] not [-1, -2]
```

Comment:
I find this comment quite hard to understand.

Reply:
I try to update it a little bit. I hope it is better now.